llvm/lib/IR/RepoTicket.cpp

```diff
@@ -116,78 +125,135 @@ static GODigestState accumulateGOFinalDigestOrInitialDigestAndGetGODigestState(
-    // the value.
-    GOHash.update('R');
-    GOHash.update(StateIt->second);
-    return true;
+    assert(NumVisited > StateIt->second);
+    // If GO is visited, mark as a Backref and use its state as the value.
+    GOHash.update(static_cast<char>(Tags::Backref));
+    GOHash.update(NumVisited - StateIt->second - 1U);
+    LLVM_DEBUG(dbgs() << "Hashing back reference to #" << StateIt->second
+                      << "\n");
+    MD5::MD5Result Digest;
+    GOHash.final(Digest);
+    return std::make_tuple(true, StateIt->second, Digest);
   }
 
-  bool Changed = false;
-  GOHash.update('T');
+  GOHash.update(static_cast<char>(Tags::GO));
   auto State = AccumulateGODigest(GO, GOHash, GOIMap);
-  Changed = Changed || State.Changed;
-  // Update the GO's hash using its contributions and dependencies.
-  auto &GOContributions = State.Contributions;
-  auto &GODependencies = State.Dependencies;
-  for (const GlobalObject *const G : llvm::concat<const GlobalObject *const>(
-           make_range(GOContributions.begin(), GOContributions.end()),
-           make_range(GODependencies.begin(), GODependencies.end()))) {
+  // Add the GO's contributions to its hash. Contributions can't form loops
+  // since 1) only global variables have contributions and 2) all contributions
+  // are functions.
+  bool Changed = State.Changed || !State.Contributions.empty();
+  for (const GlobalObject *const G : State.Contributions) {
+    assert(isa<GlobalVariable>(GO) &&
+           "Only global variables can have contributions!");
+    assert(isa<Function>(G) && "All contributions are functions!");
+    GOHash.update(GOIMap[G].InitialDigest.Bytes);
+  }
+  // Add the GO's dependencies to its hash.
+  auto LoopPoint = std::numeric_limits<size_t>::max();
+  for (const GlobalObject *const G : State.Dependencies) {
     const llvm::Function *const Fn = dyn_cast<const llvm::Function>(G);
     // if function will not be inlined and not be discarded if it is not used,
     // skip it.
     if (Fn && Fn->hasFnAttribute(Attribute::NoInline) &&
         !Fn->isDiscardableIfUnused())
       continue;
-    Changed = updateDigestUseDependenciesAndContributions(
-                  G, GOHash, Visited, GOIMap, AccumulateGODigest) ||
-              Changed;
+    const auto GState = updateDigestUseDependenciesAndContributions(
+        G, Visited, GOIMap, GOHashCache, AccumulateGODigest);
+    Changed = Changed || std::get<0>(GState);
+	// A GO which loops back to itself doesn't count as a loop.
+    if (G != GO) {
+      LoopPoint = std::min(LoopPoint, std::get<1>(GState));
+    }
+    GOHash.update(std::get<2>(GState).Bytes);
   }
+  // Encoded the final edge. Record that in the hash.
+  GOHash.update(static_cast<char>(Tags::End));
 
-  return Changed;
+  MD5::MD5Result Digest;
+  GOHash.final(Digest);
+  auto const result = std::make_tuple(Changed, LoopPoint, Digest);
+
+  if (LoopPoint > NumVisited) {
```

Comment:
I think it would be nice to pick out a variable which describes the logic for the loop here.

e.g. 
```
// Comment what this comparison is doing...
bool canCacheHash = LoopPoint > NumVisited;
if (canCacheHash) {
  GoHashCache.add(GO, Digest);
...
}
```

Reply:
I prefer not to define a variable which is only used by one time, but comments are needed. I have added some comments here.